In [83]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')

# Write query to get date, tmax, and tmin from weather
query = """
SELECT *
  FROM bronze.raw_fintech_data;
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

In [74]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')

# Write query to get date, tmax, and tmin from weather
query = """
SELECT *
  FROM silver.stg_customers;
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

In [99]:
a['last_login_date']

0       2025-12-23
1       2025-08-16
2         20251102
3       2025-06-24
4       17/03/2026
           ...    
5095    2026-03-17
5096    2025-09-09
5097    2026-04-30
5098    2025-09-21
5099    04/08/2025
Name: last_login_date, Length: 5100, dtype: str

In [77]:
df['gender'].unique()

<ArrowStringArray>
['M', 'Other', 'null', 'F', '', nan, 'N/A', 'NA']
Length: 8, dtype: str

In [70]:
sorted(b.columns)

['address',
 'avg_monthly_logins',
 'bankruptcy_flag',
 'city',
 'country',
 'credit_score',
 'credit_total_limit',
 'credit_total_used',
 'credit_utilization_pct',
 'customer_id',
 'customer_segment',
 'date_of_birth',
 'email',
 'first_name',
 'gender',
 'inquiries_6m',
 'kyc_status',
 'last_login_date',
 'last_name',
 'lat',
 'late_payments_12m',
 'load_timestamp',
 'lon',
 'mobile_app_registered',
 'nationality',
 'num_credit_accounts',
 'oldest_account_age_months',
 'paperless_statements',
 'phone_number',
 'preferred_channel',
 'push_notifications',
 'registration_date',
 'relationship_manager',
 'risk_score',
 'status',
 'web_banking_registered']

In [71]:
sorted(a.columns)


['accounts',
 'address',
 'city',
 'country',
 'credit_info.bankruptcy_flag',
 'credit_info.credit_score',
 'credit_info.currency',
 'credit_info.inquiries_6m',
 'credit_info.late_payments_12m',
 'credit_info.num_credit_accounts',
 'credit_info.oldest_account_age_months',
 'credit_info.total_limit',
 'credit_info.total_used',
 'credit_info.utilization_pct',
 'customer_id',
 'customer_segment',
 'date_of_birth',
 'digital_engagement.avg_monthly_logins',
 'digital_engagement.last_login_date',
 'digital_engagement.mobile_app_registered',
 'digital_engagement.paperless_statements',
 'digital_engagement.preferred_channel',
 'digital_engagement.push_notifications',
 'digital_engagement.web_banking_registered',
 'email',
 'first_name',
 'gender',
 'kyc_status',
 'last_name',
 'lat',
 'loans',
 'lon',
 'nationality',
 'phone_number',
 'registration_date',
 'relationship_manager',
 'risk_score',
 'status',
 'transactions']

In [56]:
pd.json_normalize(df['data']).columns

Index(['lat', 'lon', 'city', 'email', 'loans', 'gender', 'status', 'address',
       'country', 'accounts', 'last_name', 'first_name', 'kyc_status',
       'risk_score', 'customer_id', 'nationality', 'phone_number',
       'transactions', 'date_of_birth', 'customer_segment',
       'registration_date', 'relationship_manager', 'credit_info.currency',
       'credit_info.total_used', 'credit_info.total_limit',
       'credit_info.credit_score', 'credit_info.inquiries_6m',
       'credit_info.bankruptcy_flag', 'credit_info.utilization_pct',
       'credit_info.late_payments_12m', 'credit_info.num_credit_accounts',
       'credit_info.oldest_account_age_months',
       'digital_engagement.last_login_date',
       'digital_engagement.preferred_channel',
       'digital_engagement.avg_monthly_logins',
       'digital_engagement.push_notifications',
       'digital_engagement.paperless_statements',
       'digital_engagement.mobile_app_registered',
       'digital_engagement.web_banking_regis

# Exploración

In [21]:
df_exploded = pd.json_normalize(
    df['data'], 
    record_path=['transactions'], 
    meta=['customer_id']
)

In [9]:
data = pd.json_normalize(df['data'].values)

In [195]:
data.head(3)

,lat,lon,city,email,loans,gender,status,address,country,accounts,...,credit_info.late_payments_12m,credit_info.num_credit_accounts,credit_info.oldest_account_age_months,digital_engagement.last_login_date,digital_engagement.preferred_channel,digital_engagement.avg_monthly_logins,digital_engagement.push_notifications,digital_engagement.paperless_statements,digital_engagement.mobile_app_registered,digital_engagement.web_banking_registered
0,-18.551408,-69.638686,La Serna,lucas.brito@gmail.com,"[{'type': 'business', 'status': 'default', 'lo...",F,suspended,"Pasaje Bolivia 637 Interior 460, Nueva Côte d'...",CL,"[{'status': 'frozen', 'balance': 292109.25, 'c...",...,7,2,242,2025-12-23,phone,60,False,True,True,False
1,-4.476765,-68.027337,Salvador,oliver.montalbán@bol.com.br,"[{'type': 'education', 'status': 'paid_off', '...",M,inactive,"Av. carrera 5ª # 12-48 Sur, Casa 55, 054138, R...",BR,"[{'status': 'closed', 'balance': 405626.68, 'c...",...,3,6,40,2025-08-16,branch,20,False,True,False,True
2,8.009853,-73.370236,Medellin,matilda.olivé@gmail.com,"[{'type': 'education', 'status': 'delinquent',...",Other,suspended,"Dg. 9U # 6-8 Sur, Bodega 108, 258646, Tocancip...",CO,"[{'status': 'closed', 'balance': 271756.62, 'c...",...,8,1,271,20251102,phone,39,True,True,True,False


In [17]:
loan = pd.json_normalize(data['loans'].explode())

In [24]:
pd.Series(loan['principal'].unique()).to_csv("temp.csv")

In [196]:
len(data)

5100

In [197]:
df_duplicated = data[data['customer_id'].duplicated()]
len(df_duplicated)

100

In [198]:
data = data[~data['customer_id'].duplicated()]

In [199]:
data.isnull().sum().sort_values(ascending=False)

relationship_manager                         111
address                                       78
digital_engagement.avg_monthly_logins         68
gender                                        47
city                                           0
lon                                            0
lat                                            0
status                                         0
loans                                          0
email                                          0
accounts                                       0
first_name                                     0
kyc_status                                     0
risk_score                                     0
country                                        0
customer_id                                    0
nationality                                    0
transactions                                   0
phone_number                                   0
date_of_birth                                  0
customer_segment    

In [200]:
data['digital_engagement.avg_monthly_logins'].unique()

array([60, 20, 39, 6, 57, 35, 44, 12, 45, 3, 22, 53, 40, 52, 41, 1, 7, 13,
       55, 11, 23, 9, 56, 46, 16, 50, 18, 21, 5, 30, 27, 'null', 49, None,
       59, 2, 4, 26, 25, 58, 0, 38, 'N/A', 8, 28, 43, 24, 36, 51, 33, 17,
       47, 'NA', 48, 19, 34, 42, 10, 15, 29, 54, 32, '', 14, 37, 31],
      dtype=object)

In [201]:
missing_values = ['NA', 'N/A', 'null', 'nan', '',' ', None]

In [202]:

for col in data.columns:
    if data[col].isin(missing_values).any():
        print(col)


gender
address
relationship_manager
digital_engagement.avg_monthly_logins


# Reporte de Valores Nulos

| Variable | Nulos |
| :--- | :---: |
| relationship_manager | 111 |
| address | 78 |
| avg_monthly_logins | 68 |
| gender | 47 |


## relationship_manager

In [203]:
data['relationship_manager'].unique()

<StringArray>
[  'Luis Hernandez',      'Sofia Lopez', 'Valentina Torres',
     'Ana Martinez',  'Nicolas Mendoza',               'NA',
    'Daniela Rojas',     'Lucia Ortega',   'Alejandro Diaz',
   'Andres Morales',                nan,     'Maria Garcia',
     'Camila Reyes',  'Isabella Vargas', 'Sebastian Castro',
 'Carlos Rodriguez',             'null',    'Diego Ramirez',
              'N/A',                 '']
Length: 20, dtype: str

In [204]:
missing_values = ['NA', 'N/A', 'null', 'nan', '', None]

data['relationship_manager'] = (
    data['relationship_manager']
    .replace(missing_values, np.nan)
)

In [205]:
data['relationship_manager'].unique()


<StringArray>
[  'Luis Hernandez',      'Sofia Lopez', 'Valentina Torres',
     'Ana Martinez',  'Nicolas Mendoza',                nan,
    'Daniela Rojas',     'Lucia Ortega',   'Alejandro Diaz',
   'Andres Morales',     'Maria Garcia',     'Camila Reyes',
  'Isabella Vargas', 'Sebastian Castro', 'Carlos Rodriguez',
    'Diego Ramirez']
Length: 16, dtype: str

## address

In [206]:
data['address'][data['address'].isin(missing_values)].unique()

<StringArray>
['NA', 'null', '', 'N/A', nan]
Length: 5, dtype: str

In [207]:
data['address'] = (
    data['address']
    .replace(missing_values, np.nan)
)

In [208]:
data['address'][data['address'].isin(missing_values)].unique()

<StringArray>
[nan]
Length: 1, dtype: str

## digital_engagement.avg_monthly_logins

In [209]:
data['digital_engagement.avg_monthly_logins'].unique()

array([60, 20, 39, 6, 57, 35, 44, 12, 45, 3, 22, 53, 40, 52, 41, 1, 7, 13,
       55, 11, 23, 9, 56, 46, 16, 50, 18, 21, 5, 30, 27, 'null', 49, None,
       59, 2, 4, 26, 25, 58, 0, 38, 'N/A', 8, 28, 43, 24, 36, 51, 33, 17,
       47, 'NA', 48, 19, 34, 42, 10, 15, 29, 54, 32, '', 14, 37, 31],
      dtype=object)

In [210]:
data['digital_engagement.avg_monthly_logins'] = (
    data['digital_engagement.avg_monthly_logins']
    .replace(missing_values, np.nan)
)

In [211]:
data['digital_engagement.avg_monthly_logins'].unique()

array([60, 20, 39, 6, 57, 35, 44, 12, 45, 3, 22, 53, 40, 52, 41, 1, 7, 13,
       55, 11, 23, 9, 56, 46, 16, 50, 18, 21, 5, 30, 27, nan, 49, 59, 2,
       4, 26, 25, 58, 0, 38, 8, 28, 43, 24, 36, 51, 33, 17, 47, 48, 19,
       34, 42, 10, 15, 29, 54, 32, 14, 37, 31], dtype=object)

## gender

In [212]:
data['gender'].unique()

<StringArray>
['F', 'M', 'Other', '', 'N/A', 'null', 'NA', nan]
Length: 8, dtype: str

In [213]:
data['gender'] = (
    data['gender']
    .replace(missing_values, np.nan)
)

In [214]:
data['gender'].unique()


<StringArray>
['F', 'M', 'Other', nan]
Length: 4, dtype: str

## Reading again the count of null values

In [215]:
data.isnull().sum().sort_values(ascending=False)

relationship_manager                         496
address                                      393
digital_engagement.avg_monthly_logins        374
gender                                       271
city                                           0
lon                                            0
lat                                            0
status                                         0
loans                                          0
email                                          0
accounts                                       0
first_name                                     0
kyc_status                                     0
risk_score                                     0
country                                        0
customer_id                                    0
nationality                                    0
transactions                                   0
phone_number                                   0
date_of_birth                                  0
customer_segment    

# Revisión cada columna

In [216]:
pd.json_normalize(df['data']).columns

Index(['lat', 'lon', 'city', 'email', 'loans', 'gender', 'status', 'address',
       'country', 'accounts', 'last_name', 'first_name', 'kyc_status',
       'risk_score', 'customer_id', 'nationality', 'phone_number',
       'transactions', 'date_of_birth', 'customer_segment',
       'registration_date', 'relationship_manager', 'credit_info.currency',
       'credit_info.total_used', 'credit_info.total_limit',
       'credit_info.credit_score', 'credit_info.inquiries_6m',
       'credit_info.bankruptcy_flag', 'credit_info.utilization_pct',
       'credit_info.late_payments_12m', 'credit_info.num_credit_accounts',
       'credit_info.oldest_account_age_months',
       'digital_engagement.last_login_date',
       'digital_engagement.preferred_channel',
       'digital_engagement.avg_monthly_logins',
       'digital_engagement.push_notifications',
       'digital_engagement.paperless_statements',
       'digital_engagement.mobile_app_registered',
       'digital_engagement.web_banking_regis

In [217]:
import pandas as pd

def validar_coordenadas(df):
    # 1. Verificar valores nulos
    nulos_lat = df['lat'].isnull().sum()
    nulos_lon = df['lon'].isnull().sum()
    
    # 2. Definir rangos válidos
    # Latitud: -90 a 90 | Longitud: -180 a 180
    fuera_rango = df[
        (df['lat'] < -90) | (df['lat'] > 90) | 
        (df['lon'] < -180) | (df['lon'] > 180)
    ]
    
    print("--- Reporte de Validación ---")
    print(f"Valores nulos en Latitud: {nulos_lat}")
    print(f"Valores nulos en Longitud: {nulos_lon}")
    print(f"Registros fuera de rango geográfico: {len(fuera_rango)}")
    
    if not fuera_rango.empty:
        print("\nEjemplos de datos erróneos:")
        print(fuera_rango[['lat', 'lon']].head())
    else:
        print("\n✅ Todos los valores están dentro de los rangos lógicos.")
    return fuera_rango

# Ejecución
fuera = validar_coordenadas(data)

--- Reporte de Validación ---
Valores nulos en Latitud: 0
Valores nulos en Longitud: 0
Registros fuera de rango geográfico: 340

Ejemplos de datos erróneos:
           lat         lon
38   43.213416  276.536473
41 -133.848080  -71.792598
53  140.816197 -378.344231
70 -195.174113  361.343921
92 -139.702335  320.180553


## lat lon

In [218]:
def validate_coords(df):
    lat_bad = ~df['lat'].between(-90, 90)
    lon_bad = ~df['lon'].between(-180, 180)

    df['geo_valid'] = ~(lat_bad | lon_bad)

    return df
validate_coords(data)[validate_coords(data)['geo_valid'] == False]

,lat,lon,city,email,loans,gender,status,address,country,accounts,...,credit_info.num_credit_accounts,credit_info.oldest_account_age_months,digital_engagement.last_login_date,digital_engagement.preferred_channel,digital_engagement.avg_monthly_logins,digital_engagement.push_notifications,digital_engagement.paperless_statements,digital_engagement.mobile_app_registered,digital_engagement.web_banking_registered,geo_valid
38,43.213416,276.536473,Mendoza,ocampo.gomez@hotmail.com,"[{'type': 'education', 'status': 'paid_off', '...",Other,Cerrado,"Calzada Ledesma 313 097, San Eugenia los altos...",AR,"[{'status': 'closed', 'balance': 116275261.94,...",...,10,43,2025-12-05,branch,27,True,False,False,False,False
41,-133.848080,-71.792598,Buenos Aires,damián.ramos@gmail.com,"[{'type': 'personal', 'status': 'default', 'lo...",M,inactive,"Circuito Sur Olivo 135 876, San Jorge los bajo...",AR,"[{'status': 'frozen', 'balance': '237661.44 US...",...,2,116,03/02/2026,atm,22,yes,True,True,True,False
53,140.816197,-378.344231,Barranquilla,elkin.pastor@hotmail.com,"[{'type': 'auto', 'status': 'default', 'loan_i...",M,inactive,"Av. Neuquén N° 31 Local 97, Comodoro Rivadavia...",CO,"[{'status': 'closed', 'balance': 285806.07, 'c...",...,2,117,2025-08-11,branch,52,True,False,True,False,False
70,-195.174113,361.343921,Buenos Aires,cristina.véliz@gmail.com,[],NaN,active,"Pasaje Norte Candelaria 568 276, San Alfredo l...",AR,"[{'status': 'closed', 'balance': 415463.37, 'c...",...,2,52,2025-08-24,atm,52,True,True,False,True,False
92,-139.702335,320.180553,Lima,nilo.gonzalez@hotmail.com,[],M,suspended,"Continuación Sur Zambrano 860 047, San José Ma...",PE,"[{'status': 'active', 'balance': 142479.13, 'c...",...,2,175,2025-06-21,branch,57,True,False,True,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5058,-177.699488,-183.580501,Rio de Janeiro,lucía.rodrigues@yahoo.com.br..,"[{'type': 'auto', 'status': 'paid_off', 'loan_...",M,activo,"Blv. Rosario N° 66 Piso 3 Dto. 2, Viedma 8500,...",BR,"[{'status': 'closed', 'balance': 352690.34, 'c...",...,5,250,2025-05-15,atm,9,False,False,False,False,False
5061,160.844099,228.042959,Montevideo,aurelio.nava@hotmail.com,[],Other,active,"Calle Irán 620 Edif. 986 , Depto. 970, Vieja C...",UY,"[{'status': 'active', 'balance': 182535.11, 'c...",...,10,345,2025-06-25,web,47,True,False,True,True,False
5063,191.256381,29.415876,Paysandu,maría.lara@hotmail.com,"[{'type': 'personal', 'status': 'default', 'lo...",NaN,inactive,NaN,UY,"[{'status': 'closed', 'balance': 3964170.77, '...",...,7,113,05/07/2025,branch,39,True,True,False,False,False
5076,-144.065031,374.347711,Paysandu,ángela.cavalcanti@yahoo.com,"[{'type': 'personal', 'status': 'default', 'lo...",NaN,inactive,"Glorieta de Leandro Font 681 Apt. 15 , Girona,...",UY,"[{'status': 'active', 'balance': '16787556.91'...",...,10,116,2026-01-05,branch,52,False,True,False,True,False


## city

In [219]:
data['city'].str.strip().unique()

<StringArray>
[        'La Serna',         'Salvador',         'Medellin',
       'Montevideo',           'Temuco',         'Fortalza',
          'Mendoza',       'Concepcion',          'Cordoba',
         'Brasilia', 'Ciudad de Mexico',      'Guadalajara',
        'Fortaleza',            'Bogta',       'Guadalajra',
         'Cartagna',          'Brasila',            'Salto',
           'Bogota',        'Sao Paulp',            'Cusco',
          'Salvdor',       'Valparaiso',      'Barranquila',
          'Medelin',            'Saltp',          'Rosario',
            'Rivra',  'Ciudd de Mexico',     'Buenos Aires',
            'Piura',        'Maldonado',     'Barranquilla',
         'Monterry',          'Tijuana',              'Cai',
        'Cartagena',        'Sao Paulo',   'Rio de Janeiro',
         'Paysandu',         'Arequipa',        'Monterrey',
        'La Serena',             'Cali',           'Mendza',
           'Rivera',      'Buenos Aire',             'Lima',
         '

## email

In [220]:
EAI (Email Address Internationalization)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (3252061069.py, line 1)

In [ ]:
dominios_unicos = data['email'].str.split('@').str[1].str.lower().unique()

print(f"Se encontraron {len(dominios_unicos)} dominios únicos:")
print(dominios_unicos)

Se encontraron 37 dominios únicos:
['gmail.com' 'bol.com.br' 'hotmail.com' 'yahoo.com.br' 'live.com'
 'terra.com.br' 'yahoo.com' 'icloud.com' 'uol.com.br' nan 'outlook.com.br'
 'livecom' '#hotmail.com' 'gmail.com..' 'yahoo.com..' '' 'hotmailcom'
 ' yahoo.com' '#terra.com.br' 'yahoocom' ' hotmail.com' '#gmail.com'
 '#yahoo.com.br' 'terra.com.br..' 'gmailcom' 'hotmail.com..' 'yahoocombr'
 'yahoo.com.br..' 'outlook.com.br..' '#outlook.com.br' '#yahoo.com'
 'uolcombr' 'icloudcom' ' live.com' 'uol.com.br..' ' gmail.com'
 ' icloud.com']


In [ ]:
data[data['email'].str.count('@')>1].sample(5)

,lat,lon,city,email,loans,gender,status,address,country,accounts,...,credit_info.oldest_account_age_months,digital_engagement.last_login_date,digital_engagement.preferred_channel,digital_engagement.avg_monthly_logins,digital_engagement.push_notifications,digital_engagement.paperless_statements,digital_engagement.mobile_app_registered,digital_engagement.web_banking_registered,geo_valid,geo_invalid_reason
2754,-22.234010,-70.097817,Salvador,@@isaac.ruiz@yahoo.com,"[{'type': 'education', 'status': 'default', 'l...",M,closed,"Diagonal 92 # 39-5 Sur, 709729, Ovejas, Sucre",BR,"[{'status': 'active', 'balance': 493384.97, 'c...",...,345,2025-08-16,mobile,35,True,False,False,False,True,None
3471,-29.911778,-70.670263,Temuco,@@martina.velandia@yahoo.com,"[{'type': 'auto', 'status': 'delinquent', 'loa...",M,closed,Calle San Miguel de Tucumán N° 71 Torre 7 Dto....,CL,"[{'status': 'frozen', 'balance': 322994.59, 'c...",...,322,2026-02-28,web,NaN,True,True,True,True,True,None
1734,-162.127303,316.680734,Fortaleza,@@thiago.segura@uol.com.br,"[{'type': 'personal', 'status': 'delinquent', ...",F,inactive,"Av. J.J. Castelli N° 5586 Dto. 2, Ushuaia 9410...",BR,"[{'status': 'Active', 'balance': 441038.55, 'c...",...,55,2025-06-24,branch,NaN,False,false,False,True,False,both_out_of_range
4156,-17.470683,-80.367548,Arequpa,@@facundo.rodriguez@yahoo.com,[],Other,suspended,"Estação Moreira, 454, Sport Club, 93290-018 Fe...",PE,"[{'status': 'frozen', 'balance': '441430,37', ...",...,76,2026-01-30,web,51,True,false,True,True,True,None
4169,29.592524,-104.728333,Monterrey,@@alta gracia.montoya@hotmail.com,[],F,closed,"Calle de Danilo Jaén 19 Piso 0 , Zamora, 13108",MX,"[{'status': 'frozen', 'balance': 365949.21, 'c...",...,246,2025-12-14,web,52,False,True,True,False,True,None


In [ ]:
from email_validator import validate_email, EmailNotValidError

def check_email(email):
    try:
        validate_email(email, check_deliverability=False)
        return True
    except EmailNotValidError:
        return False
data[data['email'].apply(check_email)]['email']

0              lucas.brito@gmail.com
1        oliver.montalbán@bol.com.br
2            matilda.olivé@gmail.com
3       armando.villalba@hotmail.com
4           jade.camargo@hotmail.com
                    ...             
5090     mohamed.abascal@hotmail.com
5091         amador.duarte@gmail.com
5093     edelmiro.olmos@terra.com.br
5097    joaquina.martínez@bol.com.br
5098           inés.moreno@yahoo.com
Name: email, Length: 4244, dtype: str

## nationality

In [ ]:
data['nationality'].unique()

<StringArray>
['CL', 'BR', 'CO', 'UY', 'AR', 'MX', 'PE']
Length: 7, dtype: str

In [ ]:
data['phone_number'].to_csv('temp.csv')

In [ ]:
data['relationship_manager'].to_csv('temp.csv')

In [ ]:
data['credit_info.total_limit'].to_csv('temp.csv')

In [ ]:
data['digital_engagement.web_banking_registered'].unique()

array([False,  True])

In [ ]:
data['transactions'].values

array([list([{'date': '2026-04-11', 'type': 'Withdrawal', 'amount': 14039.03, 'status': 'pending', 'channel': 'pos', 'category': 'insurance', 'currency': 'USD', 'merchant': None, 'account_id': 'ACC-405300E15156', 'description': 'Acciones madrid vio trata llega debía.', 'transaction_id': 'TXN-7A7F1852B5E4'}, {'date': '05-05-2026', 'type': 'fee', 'amount': '$33593.14', 'status': 'completed', 'channel': 'mobile', 'category': 'shopping', 'currency': 'USD', 'merchant': 'Sodimac', 'account_id': 'ACC-405300E15156', 'description': 'Primeras señora habla vivir posición oficial tienen todas.', 'transaction_id': 'TXN-BBF37C53AEE4'}, {'date': '2026-03-16', 'type': 'refund', 'amount': 9307.66, 'status': 'completed', 'channel': 'pos', 'category': 'investment', 'currency': 'USD', 'merchant': 'Falabella', 'account_id': 'ACC-405300E15156', 'description': 'Ley fueron evitar poco.', 'transaction_id': 'TXN-68E6F9DE3AA7'}, {'date': '2026-04-18', 'type': 'payment', 'amount': 9451.52, 'status': 'completed', 

In [141]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')

# Write query to get date, tmax, and tmin from weather
query = """
SELECT *
  FROM silver.digital_engagement;
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

# credit_info

In [317]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')
tables = ['silver.credit_info','silver.digital_engagement','silver.stg_customers']
# Write query to get date, tmax, and tmin from weather
query = f"""
select *
from {tables[0]}
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

In [247]:
df.columns

Index(['customer_id', 'credit_score', 'utilization_pct', 'total_limit',
       'total_used', 'oldest_account_age_months', 'late_payments_12m',
       'inquiries_6m', 'bankruptcy_flag'],
      dtype='str')

In [262]:
df['credit_score'].isna().sum()

np.int64(19)

## utilization_pct

In [285]:
df['credit_score'].isna().sum()


np.int64(902)

puesto que no podemos saber si fue un error al ingresar USD de más o utilizar

In [102]:
df

,last_login_date,preferred_channel,avg_monthly_logins,push_notifications,paperless_statements,mobile_app_registered,web_banking_registered
0,2025-12-23,phone,60,False,True,True,False
1,2025-08-16,branch,20,False,True,False,True
2,20251102,phone,39,True,True,True,False
3,2025-06-24,branch,6,True,False,False,False
4,17/03/2026,mobile,57,True,True,True,False
...,...,...,...,...,...,...,...
5095,2026-03-17,atm,19,False,False,True,True
5096,2025-09-09,mobile,5,True,True,False,False
5097,2026-04-30,mobile,9,False,False,True,False
5098,2025-09-21,phone,50,False,False,False,True


## total_limit

In [292]:
df['total_limit'].isna().sum()

np.int64(0)

In [307]:
import pandas as pd

def convertible_a_entero(serie: pd.Series) -> pd.Series:
    """
    Recibe una serie de pandas y retorna una serie con los valores únicos
    que NO se pueden convertir a números flotantes.
    """
    # 1. Intentamos la conversión. Los errores se convierten en NaN
    conversion = pd.to_numeric(serie, errors='coerce')
    
    # 2. Filtramos la serie original donde la conversión falló (es NaN)
    # y usamos .unique() para no repetir los mismos errores
    valores_invalidos = serie[conversion.isna()].unique()
    
    return valores_invalidos

In [293]:
list(obtener_no_convertibles(df['total_limit']))

[]

## total_used

In [297]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')
tables = ['silver.credit_info','silver.digital_engagement','silver.stg_customers']
# Write query to get date, tmax, and tmin from weather
query = f"""
select *
from {tables[0]}
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

In [301]:
df['total_used'].values[0]

np.float64(744892.18)

In [299]:
list(obtener_no_convertibles(df['total_used']))

[]

## oldest_account_age_months

In [313]:
import pandas as pd

def no_convertible_a_entero(serie: pd.Series) -> pd.Series:
    """
    Recibe una serie de pandas y retorna una serie con los valores únicos
    que NO se pueden convertir a números flotantes.
    """
    # 1. Intentamos la conversión. Los errores se convierten en NaN
    conversion = pd.to_numeric(serie, errors='coerce')
    
    # 2. Filtramos la serie original donde la conversión falló (es NaN)
    # y usamos .unique() para no repetir los mismos errores
    valores_invalidos = serie[conversion.isna()].unique()
    
    return valores_invalidos

In [308]:
no_convertible_a_entero(df['oldest_account_age_months'])

<ArrowStringArray>
[]
Length: 0, dtype: str

## late_payments_12m

In [310]:
no_convertible_a_entero(df['late_payments_12m'])

<ArrowStringArray>
[]
Length: 0, dtype: str

## inquiries_6m

In [312]:
no_convertible_a_entero(df['inquiries_6m'])

<ArrowStringArray>
[]
Length: 0, dtype: str

## bankruptcy_flag

In [319]:
df['bankruptcy_flag']

0        True
1       False
2        True
3        True
4        True
        ...  
4995     True
4996     True
4997     True
4998    False
4999    False
Name: bankruptcy_flag, Length: 5000, dtype: bool

# digital_engagement

In [580]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')
tables = ['silver.credit_info','silver.digital_engagement','silver.stg_customers']
# Write query to get date, tmax, and tmin from weather
query = f"""
select *
from silver.customers
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

In [534]:
df['country'].unique()

<ArrowStringArray>
['BR', 'AR', 'MX', 'PE', 'UY', 'CO', 'CL']
Length: 7, dtype: str

In [583]:
df['risk_score'].isna().sum()

np.int64(0)

In [582]:
df['risk_score'].unique()

array([87.24, 19.62, 22.45, ..., 66.99, 13.9 , 28.66], shape=(3937,))

In [561]:
df['lat'].isna().sum()

np.int64(237)

In [463]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7663 entries, 0 to 7662
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   customer_id          7663 non-null   str    
 1   loan_id              7663 non-null   str    
 2   type                 7663 non-null   str    
 3   principal            7663 non-null   float64
 4   outstanding_balance  7663 non-null   float64
 5   interest_rate        7663 non-null   float64
 6   term_months          7663 non-null   float64
 7   monthly_payment      7663 non-null   float64
 8   start_date           7663 non-null   object 
 9   end_date             7663 non-null   object 
 10  status               7663 non-null   str    
 11  days_past_due        7663 non-null   float64
 12  collateral_type      6842 non-null   str    
dtypes: float64(6), object(2), str(5)
memory usage: 1.1+ MB


In [429]:
import pandas as pd

def no_convertible_a_entero(serie: pd.Series) -> pd.Series:
    """
    Recibe una serie de pandas y retorna una serie con los valores únicos
    que NO se pueden convertir a números flotantes.
    """
    # 1. Intentamos la conversión. Los errores se convierten en NaN
    conversion = pd.to_numeric(serie, errors='coerce')
    
    # 2. Filtramos la serie original donde la conversión falló (es NaN)
    # y usamos .unique() para no repetir los mismos errores
    valores_invalidos = serie[conversion.isna()].unique()
    
    return valores_invalidos
no_convertible_a_entero(df['interest_rate'])


<ArrowStringArray>
[]
Length: 0, dtype: str

In [488]:
df['credit_limit']

0             NaN
1             NaN
2             NaN
3             NaN
4             NaN
           ...   
17524         NaN
17525         NaN
17526    38031.18
17527         NaN
17528         NaN
Name: credit_limit, Length: 17529, dtype: float64

In [439]:
df['term_months'].unique()

<ArrowStringArray>
['24', '180', '120', '36', '360', '60', '12', '84', '48', '240']
Length: 10, dtype: str

In [434]:
list(df['interest_rate'].unique())

np.float64(3.01)

In [419]:
df

Index(['load_timestamp', 'customer_id', 'loan_id', 'type', 'principal',
       'outstanding_balance', 'interest_rate', 'term_months',
       'monthly_payment', 'start_date', 'end_date', 'status', 'days_past_due',
       'collateral_type'],
      dtype='str')

In [414]:
df['description'].isna().sum()

np.int64(8946)

In [411]:
list(df['description'].unique())

['Fueron paz grande.',
 'Placeat vel impedit fuga.',
 'Líder horas tipos siete victoria ella calidad.',
 'Cuando cómo nueva dijo asimismo tal peor reforma.',
 'Cuerpo fuerte que conocer.',
 'Serán nuevas soy nuevo producto sur.',
 'Sola común aire presidente saber mí a cada.',
 'Otras expresión tampoco alto armas rodríguez cuando.',
 'Septiembre saber hay cuyo interior.',
 'Importancia mismos éste madrid lucha nuestros.',
 'Veritatis quae veritatis inventore.',
 'Larga tierra fecha secretario casos durante.',
 'Otros buen sector cargo esto arte cuanto región.',
 'Corrupti eligendi veritatis inventore eaque.',
 'Lejos importancia durante pasa.',
 'Hospital mayo rafael barcelona cosas posición aquellos.',
 'Instituciones punto blanco.',
 'Ea facere alias aliquam blanditiis tempora magnam.',
 'Relaciones partidos universidad blanca igual campo había.',
 'Movimiento buscar miembros.',
 'Laborum necessitatibus cum dolores iusto sapiente.',
 'Perspiciatis veritatis unde repudiandae expedita.

In [394]:
list(df['type'].unique())

['payment',
 'withdrawal',
 'comision',
 'fee',
 'refund',
 'Transfer',
 'deposit',
 'transfer',
 'FEE',
 'Deposito',
 'transferencia',
 'DEPOSIT',
 'WITHDRAWAL',
 'reembolso',
 'Deposit',
 'Refund',
 'Fee',
 'Pago',
 'pago',
 'Retiro',
 'Payment',
 'REFUND',
 'Comision',
 'Withdrawal',
 'PAYMENT',
 'deposito',
 'TRANSFER',
 'retiro']

In [387]:
list(df['date'].unique())

['2024-09-23',
 '2025-07-06',
 '2025-12-29',
 '2025-12-20',
 '2026-03-21',
 '06/02/2026',
 '2026-02-22',
 '20260226',
 '10/09/2024',
 '04-04-2026',
 '2025-11-23',
 '2026-04-23',
 '2024-01-22',
 '2025-04-04',
 '2026-02-10',
 '2024-04-28',
 '2026-04-15',
 '07-31-2022',
 '2025-07-21',
 '2025-06-16',
 '2026-02-28',
 '20251215',
 '2026-04-03',
 '2025-11-16',
 '2026-05-05',
 '2024-12-03',
 '2025-10-05',
 '2026-04-10',
 '2024-08-25',
 '01-26-2026',
 '20251227',
 '2024-08-03',
 '20251101',
 '08-10-2022',
 '2026-04-06',
 '2024-02-16',
 '2024-06-28',
 '2026-02-07',
 '2026-04-09',
 '20240809',
 '2025-09-12',
 '2025-10-22',
 '2026-01-14',
 '2025-03-14',
 '2026-02-11',
 '2025-05-25',
 '03/05/2026',
 '2025-03-07',
 '2025-10-09',
 '2025-12-19',
 '2026-05-03',
 '20260505',
 '2026-02-05',
 '2026-02-19',
 '2023-12-16',
 '19/04/2025',
 '20250213',
 '2025-11-11',
 '02-10-2026',
 '2026-04-28',
 '04-05-2026',
 '2026-03-08',
 '2025-11-10',
 '2026-01-10',
 '02-07-2026',
 '2025-07-15',
 '2026-04-20',
 '2024-04

In [345]:
import pandas as pd

def no_convertible_a_entero(serie: pd.Series) -> pd.Series:
    """
    Recibe una serie de pandas y retorna una serie con los valores únicos
    que NO se pueden convertir a números flotantes.
    """
    # 1. Intentamos la conversión. Los errores se convierten en NaN
    conversion = pd.to_numeric(serie, errors='coerce')
    
    # 2. Filtramos la serie original donde la conversión falló (es NaN)
    # y usamos .unique() para no repetir los mismos errores
    valores_invalidos = serie[conversion.isna()].unique()
    
    return valores_invalidos

In [344]:
df.isna().sum()

customer_id                0
mobile_app_registered      0
web_banking_registered     0
last_login_date            0
avg_monthly_logins        68
preferred_channel          0
push_notifications         0
paperless_statements       0
dtype: int64

## mobile_app_registered

In [340]:
df['mobile_app_registered'].unique()

array([False,  True])

## web_banking_registered

In [341]:
df['web_banking_registered'].unique()

array([ True, False])

## last_login_date

In [342]:
list(df['last_login_date'].unique())

[datetime.date(2025, 10, 23),
 datetime.date(2025, 11, 9),
 datetime.date(2026, 4, 1),
 datetime.date(2025, 8, 20),
 datetime.date(2025, 9, 9),
 datetime.date(2025, 9, 23),
 datetime.date(2025, 10, 28),
 datetime.date(2025, 8, 29),
 datetime.date(2025, 9, 13),
 datetime.date(2025, 10, 27),
 datetime.date(2026, 1, 29),
 datetime.date(2025, 11, 13),
 datetime.date(2026, 1, 11),
 datetime.date(2025, 9, 6),
 datetime.date(2026, 2, 25),
 datetime.date(2026, 1, 9),
 datetime.date(2025, 5, 25),
 datetime.date(2026, 3, 6),
 datetime.date(2026, 1, 2),
 datetime.date(2025, 6, 20),
 datetime.date(2026, 3, 18),
 datetime.date(2025, 7, 20),
 datetime.date(2025, 12, 8),
 datetime.date(2025, 5, 22),
 datetime.date(2025, 8, 24),
 datetime.date(2025, 9, 18),
 datetime.date(2025, 10, 22),
 datetime.date(2025, 12, 29),
 datetime.date(2025, 6, 26),
 datetime.date(2026, 3, 11),
 datetime.date(2026, 4, 10),
 datetime.date(2025, 5, 13),
 datetime.date(2025, 8, 7),
 datetime.date(2025, 7, 14),
 datetime.date(

## avg_monthly_logins

In [349]:
no_convertible_a_entero(df['avg_monthly_logins'])

<ArrowStringArray>
['', 'N/A', 'NA', 'null', nan]
Length: 5, dtype: str

## preferred_channel

In [351]:
df['preferred_channel'].unique()

<ArrowStringArray>
['web', 'mobile', 'branch', 'atm', 'phone']
Length: 5, dtype: str

## push_notifications

In [353]:
df['push_notifications'].unique()

<ArrowStringArray>
['true', 'false', 'no', 'yes', '1', '0']
Length: 6, dtype: str

## paperless_statements

In [355]:
df['paperless_statements'].unique()

<ArrowStringArray>
['true', 'false', 'no', '1', '0', 'si']
Length: 6, dtype: str

# Age calculation

In [39]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')
tables = ['silver.credit_info','silver.digital_engagement','silver.stg_customers']
# Write query to get date, tmax, and tmin from weather
query = f"""
select *
from silver.stg_customers
"""

# Make a dataframe by passing query and engine to read_sql()
df = pd.read_sql(query,engine)

In [40]:
import json
import pandas as pd

# 1. Convert the JSON strings into Python dictionaries
# (We use errors='ignore' or a lambda to handle potential malformed strings safely)
dicts = df["credit_info"].apply(json.loads)

# 2. Flatten those dictionaries into a clean, new DataFrame
df_clean = pd.json_normalize(dicts)

# 3. Join it back to your original DataFrame if you need to keep other columns
df = df.join(df_clean)

In [60]:
df_clean['utilization_pct'].unique()


<StringArray>
[ '56.13',  '59.99',  '10.18',  '38.54', '$19.28',  '69.95',  '76.26',
  '68.47',  '47.02',  '22.04',
 ...
  '59.21',  '67.85',  '82.71',  '61.83',  '63.88',  '26.42',  '56.59',
  '51.54',   '4.37',  '23.45']
Length: 3966, dtype: str

In [ ]:
"SELECT DISTINCT preferred_channel FROM silver.digital_engagement

In [1]:
from sqlalchemy import create_engine
import pandas as pd
import numpy as np

# Create database engine for data.db
engine = create_engine('postgresql://qversity-admin:qversity-admin@localhost:5432/qversity')
# Write query to get date, tmax, and tmin from weather
query = f"""
select *
from silver.digital_engagement
"""

# Make a dataframe by passing query and engine to read_sql()
base = pd.read_sql(query,engine)

In [2]:
df = base

In [6]:
df['preferred_channel'].unique()

<StringArray>
['atm', 'branch', 'mobile', 'phone', 'web']
Length: 5, dtype: str

In [57]:
base['date_of_birth'].max()

datetime.date(2008, 5, 6)